server/src/main/java/com/thoughtworks/go/config/ConfigConverter.java

```diff
@@ -395,9 +395,11 @@ private PluggableSCMMaterialConfig toPluggableScmMaterialConfig(CRPluggableScmMa
             throw new ConfigConvertionException(
                     String.format("Failed to find referenced scm '%s'", id));
 
-        return new PluggableSCMMaterialConfig(toMaterialName(crPluggableScmMaterial.getName()),
-                scmConfig, crPluggableScmMaterial.getDestination(),
-                toFilter(crPluggableScmMaterial.getFilterList()), crPluggableScmMaterial.isWhitelist());
+        PluggableSCMMaterialConfig materialConfig = new PluggableSCMMaterialConfig(toMaterialName(crPluggableScmMaterial.getName()),
+            scmConfig, crPluggableScmMaterial.getDestination(),
+            toFilter(crPluggableScmMaterial.getFilterList()), crPluggableScmMaterial.isWhitelist());
+        materialConfig.setAutoUpdate(crPluggableScmMaterial.isAutoUpdate());
```

Comment:
Ah, you mean that the code between line 370 and 392 searches for an duplicated `scmConfig`? 
To avoid that we find a `scmConfig` that we then modify we should set this flag on the `scmConfig` before doing the search. 

So around line 376 add this
```java
            scmConfig.setAutoUpdate(crPluggableScmMaterial.isAutoUpdate());
```
and revert the change on lines 399-403

Reply:
Yeah, possibly directly setting the value on the scmConfig prior to constructing is clearer that you might be modifying an existing (shared) config, since the property belongs to the overall config rather than the pluggable config.

But I'll still need to check any implications. In general the way this stuff is modelled in GoCD is a bit problematic due to the de-duplication of SCMs. It's not possible to, say, have one pipeline with a material being manual and another pipeline with the same scm (url, branch) that is auto updating (polling). This causes indeterminism by design, and confusion to users. It's even more confusing for users/admins if there are a mixture of config repo and UI-created pipelines.

The workaround for this nornally is for users to use blanket `**/*` denylists in individual pipelines' consumption of a material to vary triggering but that won't stop the polling (where people have a preference for webhooks and want to reduce load on their SCM server) - only stop the triggering if certain pipelines.

So I'll want to verify that the way this change behaves is at least similar to git, svn and other SCMs in its determinism/indeterminism.

Reply:
In the method `setCommonScmMaterialMembers` (line 480) the `autoUpdate` flag is set for material of UI-created pipelines. That method is called in `toScmMaterialConfig` for each type of SCM. There is no handling of `SCM` in that method, maybe it's handled in a different way.

To verify that it doesn't mess up with the other CSMs we could write a test that...
1. have a list of SCMs contains the same material, but with `autoUpdate=true`
2. call the `toPluggableScmMaterialConfig` method with `autoUpdate=false`
3. check that there is a new `scmConfig` created and not a reused one.

Let me know if there is anything I can give any help or if you want to have a look on your own into this area. It's a lot of details, but I guess you knows the much better than me. :-)

About the workaround. My goal is to reduce the load on the git server as we have a lot of pipelines.

Reply:
Yup, this fix overall is definitely needed, reasonable and should be able to be made work. I'll try and get some time to play "in anger", and can make the minor tweaks.

Perhaps for the non-plugin ones the de-duplication happens at runtime somehow, or I am misinterpreting the "scope" of the de-duplication that is happening here.

Do you mind sharing which SCM plugin you are using/interested in here? Otherwise I'll probably validate with https://github.com/TWChennai/gocd-git-path-material-plugin since I understand it in depth.

Reply:
I'm using the [gocd-yaml-config-plugin](https://github.com/tomzo/gocd-yaml-config-plugin) and also tested the corresponding json version just to verify that the problem wasn't with the file syntax.

Reply:
Sorry, I am asking which *SCM* plugin you are using, i.e which type of custom material you are trying to configure as `auto_update: false`.